File: laravel-elasticemail-examples/app/Http/Controllers/DoubleOptinController.php
<?php

namespace App\Http\Controllers;

use ElasticEmail\Model\ContactPayload;
use ElasticEmail\Model\EmailsPayload;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

/**
 * Double opt-in.
 * 1. subscribe: contact stored with status "Transactional" (not on the list), confirmation email sent
 *    with a link carrying an HMAC of the address
 * 2. confirm: HMAC validated, contact added to ELASTICEMAIL_LIST_NAME
 * 3. webhook: alternative confirm path driven by Elastic Email click tracking
 */
class DoubleOptinController extends Controller
{
    public function subscribe(Request $request)
    {
        $request->validate([
            'email' => 'required|email',
            'name' => 'nullable|string|max:255',
        ]);

        $email = $request->email;
        $name = (string) $request->input('name', '');
        $confirmUrl = config('elasticemail.public_url') . '/api/double-optin/confirm?email=' . rawurlencode($email) . '&token=' . $this->ee->hmac($email);

        try {
            $this->ee->contacts()->contactsPost([
                new ContactPayload([
                    'email' => $email,
                    'first_name' => explode(' ', $name)[0] ?? '',
                    'status' => 'Transactional',
                ]),
            ]);

            $result = $this->ee->sendHtml(
                $email,
                'Confirm your subscription',
                view('emails.double-optin-confirm', [
                    'welcomeText' => $name !== '' ? "Welcome, {$name}!" : 'Welcome!',
                    'confirmUrl' => $confirmUrl,
                ])->render()
            );

            return response()->json([
                'success' => true,
                'message' => 'Confirmation email sent',
                'messageId' => $result->getMessageId(),
            ]);
        } catch (\Exception $e) {
            Log::error('Double opt-in subscribe failed', ['error' => $e->getMessage()]);
            return $this->fail($e);
        }
    }

    public function confirm(Request $request)
    {
        $email = (string) $request->query('email', '');
        $token = (string) $request->query('token', '');
        $listName = config('elasticemail.list_name');

        if ($email === '' || !hash_equals($this->ee->hmac($email), $token)) {
            return response()->json(['error' => 'Invalid confirmation link'], 400);
        }

        try {
            $this->ee->lists()->listsByNameContactsPost($listName, new EmailsPayload(['emails' => [$email]]));
            Log::info("Subscription confirmed: {$email} -> {$listName}");

            if ($redirect = config('elasticemail.confirm_redirect_url')) {
                return redirect()->away($redirect);
            }
            return response()->json(['confirmed' => true, 'email' => $email, 'list' => $listName]);
        } catch (\Exception $e) {
            return $this->fail($e);
        }
    }

    /**
     * Click-tracking based confirm. Register a webhook for Clicked events pointing here.
     * On status=Clicked with the confirm URL in "target", the recipient is added to the list.
     */
    public function webhook(Request $request)
    {
        if (!$this->ee->tokenOk($request->query('token'))) {
            return response()->json(['error' => 'Invalid token'], 401);
        }

        // Elastic Email validates the URL with a GET when the webhook is saved
        if ($request->isMethod('GET') && !$request->has('status')) {
            return response()->json(['ok' => true]);
        }

        $event = array_merge($request->query(), $request->post());
        $status = $this->sanitize($event['status'] ?? '');
        $target = (string) ($event['target'] ?? '');
        $recipient = $this->sanitize($event['to'] ?? '');
        $listName = config('elasticemail.list_name');

        if ($status !== 'Clicked' || !str_contains($target, '/double-optin/confirm')) {
            return response()->json(['received' => true, 'status' => $status, 'message' => 'Event ignored']);
        }

        try {
            $this->ee->lists()->listsByNameContactsPost($listName, new EmailsPayload(['emails' => [$recipient]]));
            Log::info("Subscription confirmed via click: {$recipient}");
            return response()->json(['received' => true, 'confirmed' => true, 'email' => $recipient, 'list' => $listName]);
        } catch (\Exception $e) {
            return $this->fail($e);
        }
    }
}

File: dotnet-elasticemail-examples/MvcApp/Controllers/DoubleOptinController.cs
using System.Security.Cryptography;
using System.Text;
using ElasticEmail.Api;
using ElasticEmail.Client;
using ElasticEmail.Model;
using Microsoft.AspNetCore.Mvc;

namespace MvcApp.Controllers;

[ApiController]
public class DoubleOptinController : ControllerBase
{
    private readonly EmailsApi _emailsApi;
    private readonly ContactsApi _contactsApi;
    private readonly ListsApi _listsApi;

    public DoubleOptinController(EmailsApi emailsApi, ContactsApi contactsApi, ListsApi listsApi)
    {
        _emailsApi = emailsApi;
        _contactsApi = contactsApi;
        _listsApi = listsApi;
    }

    [HttpPost("/double-optin/subscribe")]
    public async Task<IActionResult> Subscribe([FromBody] SubscribeRequest body)
    {
        var email = body.Email;
        var name = body.Name ?? "";

        if (string.IsNullOrEmpty(email))
        {
            return BadRequest(new { error = "Missing required field: email" });
        }

        var confirmUrl = $"{Ee.PublicUrl}/double-optin/confirm?email={Uri.EscapeDataString(email)}&token={Ee.Hmac(email)}";
        var greeting = string.IsNullOrEmpty(name) ? "Welcome!" : $"Welcome, {name}!";

        try
        {
            // Stored as Transactional so it receives the confirmation but no campaigns yet
            await _contactsApi.ContactsPostAsync(new List<ContactPayload>
            {
                new ContactPayload(
                    email: email,
                    firstName: name.Split(' ', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault() ?? "",
                    status: ContactStatus.Transactional),
            });

            var result = await _emailsApi.EmailsTransactionalPostAsync(new EmailTransactionalMessageData(
                recipients: new TransactionalRecipient(to: new List<string> { email }),
                content: new EmailContent(
                    from: Ee.From,
                    subject: "Confirm your subscription",
                    body: new List<BodyPart>
                    {
                        new BodyPart(BodyContentType.HTML, $"""
                            <div style="text-align: center; padding: 40px 20px; font-family: Arial, sans-serif;">
                              <h1>{greeting}</h1>
                              <p>Please confirm your subscription to our newsletter.</p>
                              <a href="{confirmUrl}" style="background-color: #18181b; color: #fff; padding: 12px 32px; border-radius: 6px; text-decoration: none; font-weight: bold; display: inline-block;">Confirm Subscription</a>
                            </div>
                            """),
                    })));

            return Ok(new { success = true, message = "Confirmation email sent", messageId = result.MessageID });
        }
        catch (ApiException e)
        {
            return Ee.ApiError(e);
        }
    }

    [HttpGet("/double-optin/confirm")]
    public async Task<IActionResult> Confirm([FromQuery] string? email, [FromQuery] string? token)
    {
        email ??= "";
        var expected = Encoding.UTF8.GetBytes(Ee.Hmac(email));
        var given = Encoding.UTF8.GetBytes(token ?? "");
        if (email.Length == 0 || !CryptographicOperations.FixedTimeEquals(expected, given))
        {
            return BadRequest(new { error = "Invalid confirmation link" });
        }

        try
        {
            await _listsApi.ListsByNameContactsPostAsync(Ee.ListName, new EmailsPayload(emails: new List<string> { email }));
            var redirectUrl = Environment.GetEnvironmentVariable("CONFIRM_REDIRECT_URL");
            if (!string.IsNullOrEmpty(redirectUrl))
            {
                return Redirect(redirectUrl);
            }
            return Ok(new { confirmed = true, email, list = Ee.ListName });
        }
        catch (ApiException e)
        {
            return Ee.ApiError(e);
        }
    }

    // Click-tracking based confirmation: create a webhook for Clicked events pointing here.
    [HttpPost("/double-optin/webhook")]
    public async Task<IActionResult> Webhook()
    {
        if (!Ee.TokenOk(Request.Query["token"]))
        {
            return Unauthorized(new { error = "Invalid token" });
        }

        var ev = await Ee.ReadEventAsync(Request);
        var status = ev.GetValueOrDefault("status");
        var target = ev.GetValueOrDefault("target") ?? "";
        var to = ev.GetValueOrDefault("to") ?? "";

        if (status != "Clicked" || !target.Contains("/double-optin/confirm"))
        {
            return Ok(new { received = true, status = Ee.Sanitize(status), message = "Event ignored" });
        }

        try
        {
            await _listsApi.ListsByNameContactsPostAsync(Ee.ListName, new EmailsPayload(emails: new List<string> { to }));
            return Ok(new { received = true, confirmed = true, email = Ee.Sanitize(to) });
        }
        catch (ApiException e)
        {
            return Ee.ApiError(e);
        }
    }
}

public record SubscribeRequest(string? Email, string? Name);
